### DISCUSSION TITLE
One state with many internal transitions, am I thinking wrong?

### DISCUSSION THREAD

**Discussion by mong8se** [statelyai/xstate#2524] [Q&A] [answered]
Let's say I have an app that has multiple sources for events.
 
Let's say for sake of argument there are these three:

1. user actions
2. a websocket connection
3. a broadcast channel
4. an SSE connection

1 is wired from a react app `send`ing in events
2-4 are invoked callback services

Any event could come from any source, and any event could go to 1-3 (4 being unidirectional)

I have a state machine that has multiple states, and based on events events from these various sources, it transitions through various states and everything is fine.

But I've also broken out a top-level parallel state, that is basically just an event loop. Just 1 state with a lot of internal transitions back to itself, so I can trigger actions for necessary side effects.

Actions like save this value to context, forward this action to the websocket, forward that action to the broadcast channel, log that this happened, etc. Just receiving canisters from one tube and putting it in another tube, a la the central bureaucracy from Futurama.

Does this make sense? Or should I eject all the tube logic from the state machine? Or am I thinking wrong and could model this differently in the machine?

**Comment by davidkpiano** [comment, marked as answer]
It does make sense... and if you think about it, even though it might seem like some fun "tube logic", at least that logic is centralized rather than being implicit throughout your app!

There's been some discussion around event buses and different actor interaction patterns (like a receptionist) that can help with these kinds of use-cases, though. Will keep thinking through those.

**Reply by mong8se** [reply]
Got it, thanks!

I became worried when the visualization was unreadable because of so many arrows looping back on top of each other. Probably no solving that.

Appreciate your time!

**Reply by davidkpiano** [reply]
Statecharts provide a lot of features to help you organize that. Can you share your code?

**Reply by mong8se** [reply]
Something like this

```javascript
const appMachineConfig = {
  id: "appMachine",
  type: "parallel",
  context: {
    ack: "",
    systemReport: {},
    object: "",
    lastMessage: "",
    lastLog: "",
  },
  invoke: [
    {
      id: "broadcastChannel",
      src: "subscribeToBroadcastChannel",
    },
    {
      id: "worker",
      src: "connectWorker",
    },
    {
      id: "chromeExtension",
      src: "connectChromeExtension",
    },
  ],
  states: {
    routing: {
      initial: "idle",
      states: {
        idle: {
          entry: "requestConfig",
          on: {
            NOTIFY_ACK: { target: "normal", actions: "recordAck" },
          },
        },
        normal: {
          on: {
            OPEN_REPORTS: "reports",
          },
        },
        reports: {
          entry: "announceReporting",
          on: {
            WINDOW_CLOSE: "normal",
            SYSTEM_INFO: {
              actions: "forwardToChromeExtension",
            },
            SYSTEM_REPORT: {
              actions: "recordSystemReport",
            },
          },
        },
      },
    },
    eventing: {
      initial: "active",
      states: {
        active: {
          on: {
            NOTIFY_CONNECTIONS: {
              actions: "recordMessage",
            },
            NOTIFY_ACK: { actions: "recordAck" },
            EVENT_OBJECTCHANGED: {
              actions: "forwardToBroadcastChannel",
            },
            LOG_INFO: {
              actions: ["recordLog", "forwardToWorker"],
            },
            LOG_ERROR: {
              actions: "forwardToWorker",
            },
            COLOR_CHANGE: {
              actions: ["updateColor", "forwardToChromeExtension"],
            },
            CONFIG_RESPONSE: {
              actions: "updateConfig",
            },
          },
        },
      },
    },
  },
};
```

**Reply by ChrisShank** [reply]
Ya this is more or less how I would go about it! Root-level invokes for the different Web APIs you are communicating with, then you just send/receive events to/from those actors! At this level of complexity, I'm not sure how much easier we can program something like this outside of using a statechart to coordinate all of this message passing!

**Reply by mong8se** [reply]
I've been trying to think about how this would scale as we get into many many more events. In many cases, the futurama bureaucracy needs less granularity than the generators/consumers of the events, so how could I process them at a higher level?

Seems there is this: https://www.w3.org/TR/scxml/#EventDescriptors

If I understand that correctly, if I had a transition denoted like this:

```javascript
  "LOG": {
              actions: ["recordLog", "forwardToWorker"],
            },
```
it would transition on any event types that are exactly LOG or prefixed by LOG, such as `{ type: "LOG.THE_THING" }` and `{ type: "LOG.THE_OTHER_THING" }`

... not sure if this is on the roadmap for xstate though.

**Reply by davidkpiano** [reply]
> ... not sure if this is on the roadmap for xstate though.

It is, and is already implemented in the v5 branch!

You would have to specify `"LOG.*": { ... }`

**Comment by ChrisShank** [comment]
This sounds like a really cool use case and it would definitely be helpful to see some code before I could comment on anything more specifically. Based on your description it sounds like you are on the right track though with invoked callbacks, ect.! I have been exploring something very similar that would make web APIs like web sockets and SSEs first class actors in `xstate`. The repo is [here](https://github.com/ChrisShank/xstate-isomorphic-behavior/tree/master/src/real-time) if you want to check it out, would love to hear your thoughts/feedback. (Also you just reminder me that I should create a wrapper for a broadcast channel!)

P.S. Great metaphor from Futurama!! 😂
